File: Stripe/STPSourceVerification.m
//
//  STPSourceVerification.m
//  Stripe
//
//  Created by Ben Guo on 1/25/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

#import "STPSourceVerification.h"
#import "STPSourceVerification+Private.h"

#import "NSDictionary+Stripe.h"

@interface STPSourceVerification ()

@property (nonatomic, nullable) NSNumber *attemptsRemaining;
@property (nonatomic) STPSourceVerificationStatus status;
@property (nonatomic, readwrite, nonnull, copy) NSDictionary *allResponseFields;

@end

@implementation STPSourceVerification

#pragma mark - STPSourceVerificationStatus

+ (NSDictionary <NSString *, NSNumber *> *)stringToStatusMapping {
    return @{
             @"pending": @(STPSourceVerificationStatusPending),
             @"succeeded": @(STPSourceVerificationStatusSucceeded),
             @"failed": @(STPSourceVerificationStatusFailed),
             };
}

+ (STPSourceVerificationStatus)statusFromString:(NSString *)string {
    NSString *key = [string lowercaseString];
    NSNumber *statusNumber = [self stringToStatusMapping][key];

    if (statusNumber != nil) {
        return (STPSourceVerificationStatus)[statusNumber integerValue];
    }

    return STPSourceVerificationStatusUnknown;
}

+ (nullable NSString *)stringFromStatus:(STPSourceVerificationStatus)status {
    return [[[self stringToStatusMapping] allKeysForObject:@(status)] firstObject];
}

#pragma mark - Description

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],

                       // Details (alphabetical)
                       [NSString stringWithFormat:@"attemptsRemaining = %@", self.attemptsRemaining],
                       [NSString stringWithFormat:@"status = %@", ([self.class stringFromStatus:self.status]) ?: @"unknown"],
                       ];

    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPAPIResponseDecodable

+ (instancetype)decodedObjectFromAPIResponse:(NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }

    // required fields
    NSString *rawStatus = [dict stp_stringForKey:@"status"];
    if (!rawStatus) {
        return nil;
    }

    STPSourceVerification *verification = [self new];
    verification.attemptsRemaining = [dict stp_numberForKey:@"attempts_remaining"];
    verification.status = [self statusFromString:rawStatus];
    verification.allResponseFields = dict;
    return verification;
}

@end

File: StripePayments/StripePayments/API Bindings/Models/Sources/STPSourceVerification.swift
//
//  STPSourceVerification.swift
//  Stripe
//
//  Created by Ben Guo on 1/25/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

import Foundation

/// Verification status types for a Source.
@objc
public enum STPSourceVerificationStatus: Int {
    /// The verification is pending.
    case pending
    /// The verification has succeeeded.
    case succeeded
    /// The verification has failed.
    case failed
    /// The state of the verification is unknown.
    case unknown
}

/// Information related to a source's verification flow.
public class STPSourceVerification: NSObject, STPAPIResponseDecodable {
    /// The number of attempts remaining to authenticate the source object with a
    /// verification code.
    @objc public private(set) var attemptsRemaining: NSNumber?
    /// The status of the verification.
    @objc public private(set) var status: STPSourceVerificationStatus = .unknown
    @objc public private(set) var allResponseFields: [AnyHashable: Any] = [:]

    // MARK: - STPSourceVerificationStatus
    class func stringToStatusMapping() -> [String: NSNumber] {
        return [
            "pending": NSNumber(value: STPSourceVerificationStatus.pending.rawValue),
            "succeeded": NSNumber(value: STPSourceVerificationStatus.succeeded.rawValue),
            "failed": NSNumber(value: STPSourceVerificationStatus.failed.rawValue),
        ]
    }

    @objc(statusFromString:)
    class func status(from string: String) -> STPSourceVerificationStatus {
        let key = string.lowercased()
        let statusNumber = self.stringToStatusMapping()[key]

        if let statusNumber = statusNumber {
            return (STPSourceVerificationStatus(rawValue: statusNumber.intValue))!
        }

        return .unknown
    }

    @objc(stringFromStatus:)
    class func string(from status: STPSourceVerificationStatus) -> String? {
        return
            (self.stringToStatusMapping() as NSDictionary).allKeys(
                for: NSNumber(value: status.rawValue)
            )
            .first as? String
    }

    // MARK: - Description
    /// :nodoc:
    @objc public override var description: String {
        let props = [
            // Object
            String(format: "%@: %p", NSStringFromClass(STPSourceVerification.self), self),
            // Details (alphabetical)
            "attemptsRemaining = \(attemptsRemaining ?? 0)",
            "status = \((STPSourceVerification.string(from: status)) ?? "unknown")",
        ]

        return "<\(props.joined(separator: "; "))>"
    }

    // MARK: - STPAPIResponseDecodable
    override required init() {
        super.init()
    }

    public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
        guard let response = response else {
            return nil
        }
        let dict = (response as NSDictionary).stp_dictionaryByRemovingNulls() as NSDictionary

        // required fields
        let rawStatus = dict.stp_string(forKey: "status")
        if rawStatus == nil {
            return nil
        }

        let verification = self.init()
        verification.attemptsRemaining = dict.stp_number(forKey: "attempts_remaining")
        verification.status = self.status(from: rawStatus ?? "")
        verification.allResponseFields = response
        return verification
    }
}
